solutions/usaco-1159.mdx

```diff
@@ -140,3 +140,197 @@ public class ConnectingTwoBarns {
 </JavaSection>
 
 </LanguageSection>
+
+## Alternative Implementation - DFS + Binary Search
+
+As the editorial mentioned, we can minimize the cost function by using binary search. Since each conncected component array is in sorted order (if we add the nodes of each connected component in increasing order), we can binary search on the sorted connected component array to find the closest field $j$ to for each field $i$ using `std::lower_bound`.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+using vl = vector<ll>;
+
+#define pb push_back
+
+const ll MAXN = 1e5;
+
+vl adj[MAXN], comps[MAXN];
+ll comp[MAXN];
+
+// DFS to find the connected components
+
+void dfs(ll cur, ll c)
+{
+	if (comp[cur] != -1)
+	{
+		return;
+	}
+	comp[cur] = c;
+	for (ll u : adj[cur])
+	{
+		dfs(u, c);
+	}
+}
+
+ll cost(ll a, ll b)
+{
+	ll dist = MAXN;
+	for (ll u : comps[a])
+	{
+
+		/*
+		 * Finds the closest field in bs' connected
+		 * component to field u and updates the minimum
+		 * distance. The closest field is found by binary
+		 * searching on a sorted array (bs' connected 
+		 * component).
+		 */
+
+		ll i = lower_bound(comps[b].begin(),
+		comps[b].end(), u) - comps[b].begin();
+		if (i > 0)
+		{
+			dist = min(dist, abs(comps[b][i - 1] - u));
+		}
+		dist = min(dist, abs(comps[b][i] - u));
+	}
+
+	/* 
+	 * Returns the minimum cost of building a path 
+	 * between the two components (aka dist^2)
+	 */
+
+	return dist * dist;
+}
+
+void solve()
+{
+	ll n, m;
+	cin >> n >> m;
+
+	// Resets our variables for each test case
+
+	for (ll i = 0; i < n; i++)
+	{
+		comp[i] = -1, adj[i].clear(), comps[i].clear();
+	}
+
+	// Read in our input data
+
+	for (ll i = 0; i < m; i++)
+	{
+		ll a, b;
+		cin >> a >> b;
+		--a, --b;
+		adj[a].pb(b), adj[b].pb(a);
+	}
+
+	/* 
+	 * Stores the number of connected components 
+	 * (set to -1 so that we don't overcount)
+	 */
+
+	ll cur = -1;
+
+	// Uses DFS to find each connected component
+
+	for (ll i = 0; i < n; i++)
+	{
+		if (comp[i] == -1)
+		{
+			dfs(i, ++cur);
+		}
+	}
+
+	/*
+	 * Add each field to its connected component.
+	 * Since we're iterating from 0...n-1, each
+	 * connected component is guaranteed to be 
+	 * in sorted order...
+	 */
+
+	for (ll i = 0; i < n; i++)
+	{
+		comps[comp[i]].pb(i);
+	}
+
+	/*
+	 * This is used to optimize our binary search. This
+	 * ensures that the sum of the component sizes = n
+	 * The initial answer is set to the cost of building
+	 * a direct path between fields 1 and N. This also
+	 * takes care of the case when fields 1 and N are in
+	 * the same connected component.
+	 */
+
+	ll res = 0;
+	if (comps[comp[0]].size() > comps[comp[n - 1]].size())
+	{
+		res = cost(comp[n - 1], comp[0]);
+	}
+	else
+	{
+		res = cost(comp[0], comp[n - 1]);
+	}
+
+	for (ll c = 1; c < cur; c++)
+	{
+		ll cost1 = 0, cost2 = 0;
+
+		/* 
+		 * Optimize our binary search for the path between 
```

Comment:
these swaps are not necessary

Reply:
Yeah, I know, but I don't know what to use :man_shrugging:.